metagraph/src/graph/graph_extensions/node_rc.hpp

```diff
@@ -0,0 +1,54 @@
+#ifndef __NODE_RC_HPP__
+#define __NODE_RC_HPP__
+
+#include <sdsl/dac_vector.hpp>
+#include <sdsl/rrr_vector.hpp>
+
+#include "graph/representation/base/sequence_graph.hpp"
+#include "common/vectors/bit_vector_adaptive.hpp"
+
+namespace mtg {
+namespace graph {
+
+// Maps each node in a PRIMARY-mode DeBruijnGraph to nodes adjacent to its reverse
+// complement. When using the index construct constructor on DBGSuccinct, or when
+// loading such an index, this stores a map from each DBGSuccinct to the BOSS nodes
+// corresponding to the reverse complements of its k-1 prefix and k-1 suffix.
+class NodeRC : public SequenceGraph::GraphExtension {
+  public:
+    using node_index = SequenceGraph::node_index;
+
+    // If construct_index is false, then an index must be loaded to take advantage
+    // of the index for call_*_from_rc calls
+    NodeRC(const DeBruijnGraph &graph, bool construct_index = false);
+
+    void call_outgoing_from_rc(node_index node, const std::function<void(node_index)> &callback) const;
+    void call_incoming_from_rc(node_index node, const std::function<void(node_index)> &callback) const;
+
+    bool load(const std::string &filename_base);
+    void serialize(const std::string &filename_base) const;
+
+    bool is_compatible(const SequenceGraph &graph, bool verbose = true) const;
+
+  private:
+    const DeBruijnGraph *graph_;
+
+    typedef bit_vector_smart Indicator;
+    typedef sdsl::dac_vector_dp<sdsl::rrr_vector<>> Mapping;
```

Comment:
Usually `sdsl::dac_vector_dp<sdsl::rrr_vector<>>` is not significantly smaller than `sdsl::dac_vector_dp<>`. Did you compare?

Reply:
On the other hand, if it's small, it's probably accessed so rarely that it doesn't contribute to the speed much, so we can use rrr too. In any case, what's the index size compared to the graph size?

Reply:
It depends on how many near-reverse complements exist in the graph. For the BIGSI graph, for example, the index takes up 8GB, compared to 15GB and 28GB for the small and normal representations of the graph.

Reply:
Comparing with and without RRR, the execution time is the same, but RRR is 1.02x smaller, so it doesn't hurt

Reply:
I'm surprised by the overall index size, you said in the meeting it's around 50% of the graph, which means it's about 1 bit per edge. So what's the density of the indicator vector?

Reply:
The `Mapping` array has an element for 2.75% of the graph indices, and this is what takes up most of the space.

Here's a comparison of the sizes for the 750 subset of BIGSI:
`bit_vector_smart` indicator alone: 17MB

full index:
`sdsl::dac_vector_dp<sdsl::rrr_vector<>>`: 84MB
`sdsl::vlc_vector<>`: 97MB
`sdsl::int_vector<>`: 272MB
`sdsl::int_vector<>` packed: 137MB

Reply:
Can you compare this for a larger graph? Say, 30 GB?

Reply:
Just take the SRA-Microbe graph, for instance.